### PR TITLE
Move onFinishCallback before span or transaction is finished

### PR DESCRIPTION
**Pull request by stefanosiano**
## :scroll: Description
moved span and transaction finish callback before they are actually finished

## :bulb: Motivation and Context
Closes https://github.com/getsentry/sentry-java/issues/3228
This fixes and aligns the behaviour among spans and transactions.

## :green_heart: How did you test it?
Unit tests

## :pencil: Checklist

- [ ] I reviewed the submitted code.
- [ ] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry/src/main/java/io/sentry/SentryTracer.java [hunk 1]
```diff
@@ -200,6 +200,24 @@ public void finish(
...
+              oldCallback.execute(span);
+            }
+
+            // Let's call the finishCallback here, when the root span has a finished date but it's
+            // not finished, yet
+            final @Nullable TransactionFinishedCallback finishedCallback =
+                transactionOptions.getTransactionFinishedCallback();
+            if (finishedCallback != null) {
+              finishedCallback.execute(this);
+            }
```
Comment (stefanosiano) [inline sentry/src/main/java/io/sentry/SentryTracer.java:218]: I just moved the finishedCallback here, but we could move everything else, too, if needed, like the performanceCollectors

File: sentry-android-core/src/main/java/io/sentry/android/core/SpanFrameMetricsCollector.java [hunk 2]
```diff
@@ -131,19 +131,34 @@ public void onSpanFinished(final @NotNull ISpan span) {
   private void captureFrameMetrics(@NotNull final ISpan span) {
     // TODO lock still required?
     synchronized (lock) {
+      String x = "";
+      for (ISpan runningSpan : runningSpans) {
+        x += runningSpan.getOperation() + " - " + runningSpan.getDescription() + ", ";
+      }
+      System.out.println(
```
Comment (markushi) [inline sentry-android-core/src/main/java/io/sentry/android/core/SpanFrameMetricsCollector.java:138]: I guess you still want to remove those right? 😅

File: sentry/src/main/java/io/sentry/Span.java [hunk 3]
```diff
@@ -37,7 +37,9 @@ public final class Span implements ISpan {
 
   private final @NotNull IHub hub;
 
-  private final @NotNull AtomicBoolean finished = new AtomicBoolean(false);
+  private boolean finished = false;
```
Comment (markushi) [inline sentry/src/main/java/io/sentry/Span.java:40]: `nit` this probably should be volatile, but then on the other hand we also don't do this for other span properties ¯\_(ツ)_/¯

**Review by markushi — APPROVED** [review]
Nice one!

File: sentry/src/main/java/io/sentry/Span.java (hunk 3 above)
Comment (stefanosiano) [reply sentry/src/main/java/io/sentry/Span.java:40]: let's keep `AtomicBoolean` and change it later if ever needed, as i don't think it really changes anything

**Pull request merged**